### PR TITLE
Handle redirects when getting download length (#1697)

### PR DESCRIPTION
**Pull request by ssoloff**
This PR addresses the lack of support for HTTP redirects when getting a download's length as described in #1697.

This PR probably could have been a one-line fix (see the original issue description).  However, I agreed with @RoiEXLab's [advice](https://github.com/triplea-game/triplea/pull/1695#issuecomment-300465349) and felt that using the Apache HTTP client instead of `HttpURLConnection` was probably a better choice going forward.  This led me to consider other functional and refactoring issues, and I kept pulling at the thread until the PR arrived in this state. 😃  I have no problem if the team thinks this PR is too big and wants me to go back to the one-liner.

There's a lot of changes here, so I'll submit a review calling out the necessary bits in more detail.

#### Functional changes
* Replaced the `HttpURLConnection`-based implementation of **all** methods in `DownloadUtils` with one based on the Apache HTTP client.  This includes the `downloadFile()` method (since renamed), as well as the `getDownloadLength()` method, which is the target of the original issue.
* Model the download length as a `Long` instead of an `Integer` to match the underlying APIs.  This seemingly simple change is actually responsible for a good portion of the modified lines in this PR as mapping the `Long` range of the download length to the `Integer` range of the progress bar required a fundamental change in how the progress bar range was defined.
* Ensure the download length cache is thread safe, as `getDownloadLength()` is being called on multiple threads.
* Use a HEAD request instead of a GET request when getting the download length.
* Prevent the download length from being requested multiple times when a new item is selected in the UI.
* Display an indeterminate progress bar until a) the total length is known and b) the first byte is received when downloading a file (instead of leaving it at "0%" during that period, which can be for the entire download if the host does not provide download length for some reason).

#### Functional issues for review
* As stated above, it may have been overkill to replace the `downloadFile()` implementation with the Apache HTTP client because it currently is not exhibiting any known problems and already supports redirects.
* The indeterminate progress bar may not be everyone's cup of tea.  About 10-15% of the time during my testing, GitHub would not provide a `Content-Length` header (it would switch to chunked transfer encoding), and thus my progress bar would sit at 0% for the entire download and then jump to 100% when complete.  The indeterminate animation at least provides a visual cue to the user that "something" is happening.
* @RoiEXLab, I'd appreciate any general feedback you may have on my usage of the Apache HTTP client library.  Please call out anything that may not be idiomatic.

#### Refactoring changes
* I reduced method and type accessibility, where appropriate, to the minimum required for several types that I changed.
* I removed any members in the classes I touched that were no longer used.
* I switched all `DownloadUtils` URI parameters from `URL` to `String`.  We already store URIs as strings internally, and the Apache HTTP client API seems to prefer strings (or `URI`s which would have required yet another transformation).  This allowed the removal of a few adapter methods that simply convert `String` to `URL`.
* I renamed the `DownloadUtils#downloadFile()` method to `downloadToFile()`.

#### Refactoring issues for review
None.

#### Other changes
* I added the [catch-exception](https://github.com/Codearte/catch-exception/tree/2.x) library to the `testCompile` classpath to make it easier to test for expected exceptions.  I've been using it for about a year as it's much more elegant than the JUnit try/catch idiom (amid the well-known issues of false positives with `@Test(expected=...)` and the `ExpectedException` rule).

#### Other issues for review
* Please advise if the addition of catch-exception to the project is undesirable, and I'll swap it out for one of the native JUnit solutions.

#### Testing
I manually tested the UI to ensure download lengths were properly displayed and that downloads (including concurrent downloads) weren't broken in general (I tested the majority of maps, skins, and tools among all categories).  I specifically ensured the SourceForge maps that were redirected now had their download lengths displayed correctly and no errors were reported on the console.

As I noted above, I observed during my testing that about 10-15% of the time GitHub did not include a `Content-Length` header in the response (because it chose to use chunked transfer encoding).  This results in the inability to display meaningful progress information as the total length is unknown.  I do not believe this is a regression because I have seen this behavior (progress jumping from 0% to 100%) in previous releases.

There was one additional issue I encountered during testing that I cannot explain.  Very rarely (less than 5%), GET and HEAD requests to SourceForge fail due to an SSL/TLS handshake error:

```
failed to get download length for 'http://downloads.sourceforge.net/project/tripleamaps/developer%20resources/map%20makers/Map_Making_Tutorial_Map.zip'
javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
	at sun.security.ssl.Alerts.getSSLException(Alerts.java:192)
	at sun.security.ssl.Alerts.getSSLException(Alerts.java:154)
	at sun.security.ssl.SSLSocketImpl.recvAlert(SSLSocketImpl.java:2023)
	<. . . snip . . .>
	at games.strategy.engine.framework.map.download.DownloadMapsWindow.createLabelText(DownloadMapsWindow.java:307)
	at games.strategy.engine.framework.map.download.DownloadMapsWindow.lambda$11(DownloadMapsWindow.java:291)
	at java.lang.Thread.run(Thread.java:745)
```

There is no visible consequence for the end user (this condition is treated as if the download length is simply not available).  However, the above exception is logged quietly.  I don't remember ever seeing this problem in the legacy `HttpURLConnection` implementation.  Because it happens so rarely, I'm guessing it only occurs when SourceForge redirects to a particular mirror that may have an expired or otherwise invalid certificate.  If that's the case, then my guess is the default `HttpURLConnection` SSL/TLS configuration is more forgiving of invalid certificates than the default Apache HTTP client implementation.

If the team feels the rare possibility of the presence of this exception in the log (even quietly) is unacceptable, I'll try to track down the root cause of the problem.

Otherwise, I'd appreciate someone else testing this PR to ensure the Download Maps window is behaving acceptably, especially from another machine with different network characteristics than mine.

### CODE REVIEW

File: src/main/java/games/strategy/engine/framework/map/download/DownloadMapsWindow.java [hunk 1]
```diff
@@ -265,6 +265,10 @@ private static ListSelectionListener createDescriptionPanelUpdatingSelectionList
       final JList<String> gamesList, final List<DownloadFileDescription> maps, final MapAction action,
       final JLabel mapSizeLabelToUpdate) {
     return e -> {
+      if (e.getValueIsAdjusting()) {
+        return;
+      }
+
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/framework/map/download/DownloadMapsWindow.java]: A `JList` model may fire a chain of `ListSelectionEvent`s to support transition states that may be rendered during, for example, drag-and-drop.  The last event in the chain will return `false` from `getValueIsAdjusting()` and represents the final state of the list selection change.  This is the only event we should be processing to ensure multiple requests to get download length for the same URI are not fired concurrently.

File: src/main/java/games/strategy/engine/framework/map/download/DownloadMapsWindow.java [hunk 2]
```diff
@@ -294,35 +298,33 @@ private static void updateMapUrlAndSizeLabel(final DownloadFileDescription map,
...
   private static String createLabelText(final MapAction action, final DownloadFileDescription map) {
-    final String DOUBLE_SPACE = "&nbsp;&nbsp;";
+    final String doubleSpace = "&nbsp;&nbsp;";
 
-    final long mapSize;
-    String labelText = "<html>" + map.getMapName() + DOUBLE_SPACE + " v" + map.getVersion();
+    final StringBuilder sb = new StringBuilder();
+    sb.append("<html>" + map.getMapName() + doubleSpace + " v" + map.getVersion());
 
+    final Optional<Long> mapSize;
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/framework/map/download/DownloadMapsWindow.java]: I changed `mapSize` from `long` to `Optional<Long>` here to get rid of the smell of using the magic `-1` constant.  Since we were using `Optional<Long>` already, it seemed like a no-brainer.

File: src/main/java/games/strategy/engine/framework/map/download/DownloadMapsWindow.java [hunk 3]
```diff
@@ -294,35 +298,33 @@ private static void updateMapUrlAndSizeLabel(final DownloadFileDescription map,
   }
 
   private static String createLabelText(final MapAction action, final DownloadFileDescription map) {
-    final String DOUBLE_SPACE = "&nbsp;&nbsp;";
+    final String doubleSpace = "&nbsp;&nbsp;";
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/framework/map/download/DownloadMapsWindow.java]: A Checkstyle violation fix while I was here.

File: src/main/java/games/strategy/engine/framework/map/download/DownloadMapsWindow.java [hunk 4]
```diff
@@ -294,35 +298,33 @@ private static void updateMapUrlAndSizeLabel(final DownloadFileDescription map,
...
 
-    final long mapSize;
-    String labelText = "<html>" + map.getMapName() + DOUBLE_SPACE + " v" + map.getVersion();
+    final StringBuilder sb = new StringBuilder();
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/framework/map/download/DownloadMapsWindow.java]: It seemed cleaner to use a `StringBuilder` here instead of numerous string concatenations.

File: src/main/java/games/strategy/engine/framework/map/download/DownloadUtils.java [hunk 5]
```diff
@@ -1,100 +1,155 @@
...
+
+import com.google.common.annotations.VisibleForTesting;
 
 import games.strategy.debug.ClientLogger;
+import games.strategy.engine.framework.system.HttpProxy;
 
-public class DownloadUtils {
+public final class DownloadUtils {
+  @VisibleForTesting
+  static final ConcurrentMap<String, Long> downloadLengthsByUri = new ConcurrentHashMap<>();
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/framework/map/download/DownloadUtils.java]: Switched to `ConcurrentHashMap` to ensure cache access is thread safe (`getDownloadLength()` is called from multiple threads).

File: src/main/java/games/strategy/engine/framework/map/download/FileSizeWatcher.java [hunk 6]
```diff
@@ -10,26 +10,25 @@
...
  * consumer.
  */
-public class FileSizeWatcher {
-
+final class FileSizeWatcher {
   private final File fileToWatch;
-  private final Consumer<Integer> progressListener;
-  private boolean stop = false;
+  private final Consumer<Long> progressListener;
+  private volatile boolean stop = false;
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/framework/map/download/FileSizeWatcher.java]: Added `volatile` here because I noted that `stop` is accessed from multiple threads.  Probably not broken because the Java memory model is eventually consistent, but no harm in correctness.

File: src/main/java/games/strategy/engine/framework/map/download/MapDownloadProgressListener.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+import javax.swing.SwingUtilities;
+
+/**
+ * A listener of map download progress events that updates the associated controls in the UI.
+ *
+ * <p>
+ * Instances of this class are thread safe.
+ * </p>
+ */
+final class MapDownloadProgressListener {
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/framework/map/download/MapDownloadProgressListener.java]: This class was added when I changed how download length was modeled from `Integer` to `Long`.

Previously, the progress bar's maximum range was set to the download length.  When I switched the download length type to `Long`, that value was no longer compatible without casting, and casting just felt dirty (even though I doubt we'll ever have a download larger than 2 GiB :smiley:).

The solution I chose was to fix the progress bar range to [0,100] and to have the "progress updater" (this class) map the (`Long`) download progress into this (`Integer`) range (see implementation below in `downloadUpdated()`).

This class is also where all the indeterminate progress bar stuff is managed.

File: src/main/java/games/strategy/engine/framework/map/download/MapDownloadProgressListener.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+  void downloadStarted() {
+    SwingUtilities.invokeLater(() -> {
+      progressBar.setMinimum(MIN_PROGRESS_VALUE);
+      progressBar.setMaximum(MAX_PROGRESS_VALUE);
+      progressBar.setIndeterminate(true);
+      progressBar.setStringPainted(false);
+    });
+
+    downloadLength = DownloadUtils.getDownloadLength(uri);
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/framework/map/download/MapDownloadProgressListener.java]: `downloadStarted()` is always called from a dedicated background thread, so it is acceptable to call a potentially long-running network operation (`DownloadUtils#getDownloadLength()`) here.

File: src/main/java/games/strategy/engine/framework/map/download/DownloadMapsWindow.java (hunk 1 above)
Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/framework/map/download/DownloadMapsWindow.java]: I'd enclose the following code in the opposite if statement instead.
Looks better IMO

File: src/main/java/games/strategy/engine/framework/map/download/DownloadMapsWindow.java (hunk 4 above)
Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/framework/map/download/DownloadMapsWindow.java]: True, but I'd split up this line into multiple append function calls e.g. `builder.append("<html>").append(map.getMapName())....` to be consistent

File: src/main/java/games/strategy/engine/framework/map/download/DownloadMapsWindow.java [hunk 9]
```diff
@@ -294,35 +298,33 @@ private static void updateMapUrlAndSizeLabel(final DownloadFileDescription map,
...
-      mapSize = map.getInstallLocation().length();
-    }
-
-    if (mapSize > 0) {
-      labelText += DOUBLE_SPACE + " (" + createSizeLabel(mapSize) + ")";
+      mapSize = Optional.of(map.getInstallLocation().length());
     }
+    mapSize.ifPresent(size -> {
+      sb.append(doubleSpace + " (" + createSizeLabel(size) + ")");
+    });
```
Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/framework/map/download/DownloadMapsWindow.java]: This could either be shrunken to a single line (without curly brackets and semicolon) or split into multiple append calls

File: src/main/java/games/strategy/engine/framework/map/download/DownloadMapsWindow.java [hunk 10]
```diff
@@ -294,35 +298,33 @@ private static void updateMapUrlAndSizeLabel(final DownloadFileDescription map,
...
 
-    labelText += "<br/>";
+    sb.append("<br/>");
```
Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/framework/map/download/DownloadMapsWindow.java]: If we want to stick with html5 conventions/recommendations we can change this to just `<br>`

File: src/test/java/games/strategy/engine/framework/map/download/DownloadUtilsGetDownloadLengthFromHostTest.java [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+  }
+
+  @Test
+  public void shouldThrowExceptionWhenContentLengthHeaderValueIsNotNumber() throws Exception {
+    when(contentLengthHeader.getValue()).thenReturn("value");
+
+    catchException(() -> getDownloadLengthFromHost());
+
+    assertThat(caughtException(), is(instanceOf(IOException.class)));
+    assertThat(caughtException().getCause(), is(instanceOf(NumberFormatException.class)));
```
Comment (RoiEXLab) [inline src/test/java/games/strategy/engine/framework/map/download/DownloadUtilsGetDownloadLengthFromHostTest.java]: I'd ad a TODO comment, to remind us to update this code when moving to junit 5 since it brings support for test cases like this

File: src/main/java/games/strategy/engine/framework/map/download/DownloadMapsWindow.java (hunk 1 above)
Comment (ssoloff) [inline src/main/java/games/strategy/engine/framework/map/download/DownloadMapsWindow.java]: I view this check as guarding the "normal path of execution" per [Replace Nested Conditional with Guard Clauses](https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html), which I'm really OCD about (increased nesting and such). :smiley:

However, to be consistent, I could have done the same thing a few lines below when checking `somethingIsSelected`.  Since I didn't modify that line, I'll invert the condition on the line I added and nest it as you suggested.  But I bet that negative assertion is going to make it hard to fall asleep tonight. :smiley:

File: src/main/java/games/strategy/engine/framework/map/download/DownloadMapsWindow.java (hunk 9 above)
Comment (ssoloff) [inline src/main/java/games/strategy/engine/framework/map/download/DownloadMapsWindow.java]: Yes, you're right, I shouldn't have been lazy and just replaced all string concatenation at the same time.  Will do in a separate commit.

File: src/test/java/games/strategy/engine/framework/map/download/DownloadUtilsGetDownloadLengthFromHostTest.java (hunk 11 above)
Comment (ssoloff) [inline src/test/java/games/strategy/engine/framework/map/download/DownloadUtilsGetDownloadLengthFromHostTest.java]: I'd prefer not to add one more TODO to the list of 250 already in the code. :smiley:  In my experience, I find that they never get resolved if not addressed relative soon after they are introduced.  In addition, which occurrence of this usage is the best to tag with the TODO?  What if that one test gets deleted and the TODO is lost because someone is too lazy to duplicate it in another location?

Assuming the committers agree to merging this PR with catch-exception intact, I'd propose we create a new issue to track your recommendation.  Are there existing JUnit 5 migration tasks we would want to perform?  Would it be beneficial to combine them all into a single checklist issue?

**Comment by ssoloff** [comment]
> is it possible to download maps with the download window being closed?

I'm not sure what use case you are envisioning, @RoiEXLab.  Are you talking about starting downloads from the Download Maps window but being able to close the window before they are complete?  Or are you talking about starting downloads independent of the Download Maps window?

**Comment by RoiEXLab** [comment]
@ssoloff I'm talking about both.
I believe it would be the best to just be able to open the download window start some downloads, have them continue in the background, and if you open the download window again the progress bars just adjust to the current progress.

**Comment by ssoloff** [comment]
@RoiEXLab 

> I believe it would be the best to just be able to open the download window start some downloads, have them continue in the background, and if you open the download window again the progress bars just adjust to the current progress.

Definitely outside the scope of this PR, but it sounds doable.  From my limited time with the Download Maps window, I believe the biggest obstacle would be to persist some state keeping track of in-flight downloads so the UI could be restored when re-opening the window.  The downloads themselves are started in background threads, so they should keep running when the window is closed (I have not confirmed this).  The progress reporting is all based on the local file size, so as long as the total size was persisted as previously mentioned, you should be able to reconstruct the UI state appropriately.

> I'm talking about both.

For the case outside of the Download Maps window, `DownloadUtils#downloadToFile()` does the heavy lifting.  However, the `DownloadFile` class adds a few features on top of it:

* Downloads to a temporary file and then copies it to the user's downloaded maps folder only after the download completes successfully.
* Creates the map metadata (e.g. `<repoName>-<branchName>.zip.properties`).

**Comment by ron-murhammer** [comment]
@ssoloff Looks like this needs rebased.

**Comment by ssoloff** [comment]
Rebased against `master` at 1cd6290 (1.9.0.0.4324).

**Comment by RoiEXLab** [comment]
@ssoloff I rewrote the DownloadCoordinator class based on your branch: https://gist.github.com/RoiEXLab/60234738504612b25eb6da1d6a6a52eb
It should reduce uneccessary overhead as no loop is used, but the downloads invoke a check once they are complete.
The important part: It makes the DownloadFile object important. DownloadCoordinator basically queues those objects and invokes the download procedure.
If we modify `DownloadFile` to be more dynamic, to allow us to swap updateListeners we can easily decouple the downloads from the UI. Fetching active downloads? No problem, there's a method for that!

**Comment by ssoloff** [comment]
Looks good, @RoiEXLab!  I like that the download logic is becoming less spread out over multiple classes.

Just wanted to point out some things related to synchronization in the new `DownloadCoordinator` class:

* Synchronizing on `this` (and by extension, marking a method `synchronized`) is usually not recommended as it leaks implementation details outside the class (at a minimum, locking on `this` should be documented because it's effectively API).  It's preferable to synchronize on internal objects, as you did with `pendingDownloads`.
* Because multiple locks are involved, I would recommend adding a Javadoc for each field documenting what lock needs to be held when accessing the field.  If Guava or Apache Commons has something similar to Brian Goetz's `@GuardedBy` annotation, that would be even better.  (Otherwise, you could just add `net.jcip:jcip-annotations` as a dependency if the team agrees it's beneficial.)
* Possible issue: In the completion listener, you are synchronizing on `this` to access `terminatedDownloads` but then you synchronize on `terminatedDownloads` itself within `pollTerminatedDownloads()`.
* Possible issue: The `getActiveDownloads()` and `getPendingDownloads()` methods probably require synchronization while the copy is being made.
* Possible issue: Some methods appear to require acquiring multiple locks, but in actuality they do not.  For example, `cancelDownloads()` accesses `pendingDownloads` while synchronizing on `this`, but in `accept()`, `pendingDownloads` is accessed while synchronizing on `pendingDownloads`.  I interpreted that to mean I need to synchronize on `pendingDownloads` in `cancelDwonloads()`, as well.

**Comment by RoiEXLab** [comment]
@ssoloff I updated the class to avoid synchronized methods and synchronised this

**Comment by ssoloff** [comment]
Looks much better, @RoiEXLab.  However, note that `terminatedDownloads` is being manipulated in `cancelDownloads()` without acquiring the `terminatedDownloads` lock.

It's not obvious to me, so I'll just ask... is there a reason you are not just using a single lock for concurrency control for the entire instance?  Acquiring multiple locks always makes my head spin because then I have to be concerned about the order in which they are taken everywhere. :smiley:

**Comment by ssoloff** [comment]
Note that the SSL/TLS handshake error was reported by a user on the [forum](https://forums.triplea-game.org/topic/121/unable-to-download-mapmaking-tutorial-map) using an existing release.  The error message from `HttpURLConnection` is a bit more descriptive:

```
javax.net.ssl.SSLHandshakeException: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
at sun.security.ssl.Alerts.getSSLException(Unknown Source)
at sun.security.ssl.SSLSocketImpl.fatal(Unknown Source)
at sun.security.ssl.Handshaker.fatalSE(Unknown Source)
at sun.security.ssl.Handshaker.fatalSE(Unknown Source)
at sun.security.ssl.ClientHandshaker.serverCertificate(Unknown Source)
at sun.security.ssl.ClientHandshaker.processMessage(Unknown Source)
at sun.security.ssl.Handshaker.processLoop(Unknown Source)
at sun.security.ssl.Handshaker.process_record(Unknown Source)
at sun.security.ssl.SSLSocketImpl.readRecord(Unknown Source)
at sun.security.ssl.SSLSocketImpl.performInitialHandshake(Unknown Source)
at sun.security.ssl.SSLSocketImpl.startHandshake(Unknown Source)
at sun.security.ssl.SSLSocketImpl.startHandshake(Unknown Source)
at sun.net.www.protocol.https.HttpsClient.afterConnect(Unknown Source)
at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(Unknown Source)
at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(Unknown Source)
at sun.net.www.protocol.http.HttpURLConnection.getInputStream(Unknown Source)
at sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(Unknown Source)
at java.net.URL.openStream(Unknown Source)
at games.strategy.engine.framework.map.download.DownloadUtils.downloadFile(DownloadUtils.java:68)
at games.strategy.engine.framework.map.download.DownloadFile.lambda$createDownloadThread$127(DownloadFile.java:71)
at java.lang.Thread.run(Unknown Source)
```

This confirms the error is not due to the change of HTTP client in this PR.  I still think it is one SourceForge mirror with an invalid SSL/TLS certificate, which explains why it only happens intermittently.

**Comment by RoiEXLab** [comment]
@ssoloff You mean by making the INSTANCE constant volatile? 
I wanted to have a (in Germany we call this "Idiotensicher") "fool safe" object, which could have multiple instances if we wanted to.
This way nobody has to worry about breaking anything, especially because the finish listener of every download thread calls this object again.

**Comment by ssoloff** [comment]
@RoiEXLab Your philosophy is sound, and there's no need to make `INSTANCE` `volatile` as it's `final` (in fact, it's a compiler error to mark a variable both `final` and `volatile`).  It might be easier to explain what I'm talking about with an example.  I [forked your gist](https://gist.github.com/ssoloff/056dc36f35fbee1b5d39528bb5effa53) and modified it based on some of the points I mentioned above.

Some highlights:

* Uses a single lock to protect all resources in a `DownloadCoordinator` instance.  Multiple locks protecting multiple resources are only needed if the resources are independent and you need to maximize throughput (i.e. avoid contention where possible among concurrent threads).  My impression is that the state of `DownloadCoordinator` should be modified atomically, and there probably isn't a need to (at least initially) worry about performance.
* I added the JCIP concurrency annotations (`@ThreadSafe` and `@GuardedBy`) just to make it clear what's going on.  They don't actually do anything--they're simply for documentation.
* The `@GuardedBy` annotations on the instance fields inform a future maintainer that they must acquire the named lock before accessing/mutating the annotated field.
* I modified `updateQueue()` to not acquire the lock but rather require that it be acquired before the method is invoked.  A general good practice is to not acquire locks in private methods and to assume that the caller in an API method (i.e. public, protected, or package-private) acquires the appropriate lock (hence the presence of the assertion to sanity check this assumption).  The `@GuardedBy` annotation is again used to inform a future maintainer that the named lock must be acquired before calling this method.
* I added some CAUTION comments around the calls to `DownloadFile#cancelDownload()`.  In general, you should be concerned about calling out to code outside a class while you are holding a lock because that code could do something crazy that might eventually call back into the originating instance on a different thread, thus causing a deadlock.  In this case, `DownloadFile#cancelDownload()` just does some simple state manipulation and appears to be safe to call while the lock is held.  But if there's ever a way to defer these kinds of external calls until after you release the lock (e.g. buffering them in a collection instead of invoking them immediately), it's worth pursuing.
* Now that there is a single lock, I made the entire `getDownloads()` method atomic by holding the lock for the duration of the method.  Acquiring and releasing the lock three times within this method could lead to some very difficult to debug problems due to thread interleaving.  Again, I would revisit this if performance was a concern, but I don't think it is at this point.  And if I were being truly pedantic, I would have private versions of `getPendingDownloads()` and `pollTerminatedDownloads()` that do not acquire the lock, and call those versions instead to avoid re-entrancy on the lock.

Note that I only compiled this code but didn't actually test it, so there could be a runtime error lurking in there someplace.

**Comment by RoiEXLab** [comment]
@ssoloff Ah, that's what you meant...
Yeah, this is easier and should work just as good as my previous version!

**Comment by ssoloff** [comment]
@RoiEXLab I didn't receive a notification about your comment on my gist.  Just in case you aren't receiving notifications either, this is to let you know that I've responded to your question in the gist.

**Comment by RoiEXLab** [comment]
@ssoloff Awesome!
No, I didn't recieve a notification either.
I hope this PR is merged quickly so we can discuss your implementation in detail!

**Comment by ssoloff** [comment]
@ron-murhammer @DanVanAtta Bumping...

I believe @RoiEXLab has independently tested this change on his forked branch for improving the Download Maps window.  Also, the concern about the intermittent SSL/TLS handshake error I observed has been [confirmed](https://github.com/triplea-game/triplea/pull/1718#issuecomment-304416497) to exist prior to this PR, so it is not a regression.

So it just remains for a committer to approve the overall code changes.  Please advise if any of the issues for review I have identified in the description need to be addressed.

**Comment by ssoloff** [comment]
Rebased against `master` at b1af04e (1.9.0.0.4423).

**Comment by RoiEXLab** [comment]
@ssoloff Are you going to create the PR with your improved map download code now or should I do it?

**Comment by ssoloff** [comment]
@DanVanAtta Agreed that a 1:1 correspondence between production and test files is ideal.  However, in this case, I created the separate fixtures simply to share setup and utilities across multiple tests for each method under test.  I didn't want to have a gigantic fixture that had fields or methods that were only used by a subset of the tests.  I'm not sure `DownloadUtils` has reached the point where it no longer has a single responsibility.

I'm used to test frameworks in other languages that allow arbitrarily nested fixtures, so that I can have a file with a top-level fixture for the class and one nested fixture (if needed) for each method or feature.  Basically, just using the top-level fixture as a namespace for the other fixtures.

I've tried to use nested fixtures in JUnit in the past but never seemed to be able to get it to work, so I just assumed it was not supported.  I just did a search and found this [post](https://stackoverflow.com/questions/8758294/test-cases-in-inner-classes-with-junit).  There's some suggestions in there that I know I didn't try before, so I'll give it another shot and submit a PR if successful.

**Comment by ssoloff** [comment]
@RoiEXLab I was under the impression you were already working on improving this code.  I posted that gist simply because I felt obligated to after drowning you in such a long [comment](https://gist.github.com/ssoloff/056dc36f35fbee1b5d39528bb5effa53#gistcomment-2108045). :smiley:

Please feel free to use that code in your own PR.  I'll be happy to discuss what you ultimately come up with when it's ready.

**Comment by RoiEXLab** [comment]
@ssoloff
I was actually waiting for this PR to be merged before actively working on it in order to avoid any merge conlicts ^^
I will create a PR soon!

**Comment by ssoloff** [comment]
@RoiEXLab Here is the [patch](https://github.com/triplea-game/triplea/files/1039414/roiexlab-download-coordinator.diff.txt) I promised you but could not attach to my gist.  Just in case I screwed up copy-pasting code into the gist...

**Comment by RoiEXLab** [comment]
@ssoloff I pushed the download maps window to a branch:
https://github.com/RoiEXLab/triplea/tree/decouple-download
Everything works as expected, but the GUI has a glitch when starting a download, closing the window, reopening the window and starting another download.
The percent numbers are flickering between 2 states, and when reopening the window again some progress bars disappear.
Since you created the handler code, you might have a better understanding of what's going on!
I assume that a listener is not properly un-registered at some point and therefore creating the "conflicts"

**Comment by ssoloff** [comment]
@RoiEXLab NP, I'll take a look in a few hours and see if I can figure out what's going on.  I suspect you're right about the cause because I remember observing similar behavior when I first wrote it, albeit it occurred immediately upon reopen--it didn't require starting another download.  I fixed that case but probably didn't consider this one.

In the meantime, let me give you something to think about...  Since we first discussed this, I was thinking it may be more appropriate to move the `DownloadCoordinator` instance to `ClientContext` instead of hosting it within `DownloadCoordinator` itself.  `ClientContext` seems to be the preferred location for this type of service.  What do you think?
